파이썬에도 자바의 try with resource와 같은 기능을 하는것이 있다.
그것이 바로 with as 문이다.

자바에서도 try with resource를 사용하려면 해당 객체가 Closeable을 implements 해야만 했다.
파이썬에도 똑같이 with as 문에서 사용하는 객체가 __enter__()와 __exit_() 함수가 선언되어 있어야지만 with as 문을 사용할수가 잇다.
만약 __enter__()와 __exit__() 메소드가 구현되어있지 않은 클래스 객체가 with as 문에서 사용이 된다면 TypeError가 발생한다.

그럼 예제코드를 통해 사용법을 한번 봐보자.

class MyResource:
    def __init__(self):
        print("Resource created.")

    def __enter__(self):
        print("Resource entered.")
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        print("Resource exited.")

# Using the resource in a with statement
with MyResource() as resource:
    print("Using resource.")
    
위 코드를 보면 __init__() 이라는것이 매직메소드로 바로 객체 생성이 되자마자 호출되고
그다음에 with문 안으로 들어갔으니 __enter__()가 호출이 된다. 그리고 with 문 안에 명령문이 실행이 되고
명령문들이 모두 실행이 되면 __exit__()을 호출하게 된다.

따라서 결과는 아래와 같이 나오게 된다.

Resource created.
Resource entered.
Using resource.
Resource exited.
